File: AdSwitcher-iOS/AdSwitcher/AdSwitcher/AdSwitcherNativeAd.m
//
//  AdSwitcherNativeAd.m
//  AdSwitcher
//
//  Created by tkyaji on 2016/10/19.
//  Copyright © 2016年 adwitcher. All rights reserved.
//

#import "AdSwitcherNativeAd.h"

@implementation AdSwitcherNativeAd {
    NSMutableDictionary<NSString *, NSObject<NativeAdAdapter> *> *_adapterCacheDict;
    NSMutableDictionary<NSString *, AdConfig *> *_adConfigDict;
    AdSelector *_adSelector;
    
    NSObject<NativeAdAdapter> *_selectedAdapter;
    AdConfig *_selectedAdConfig;
    BOOL _loading;
    
    nativeAdReceived _nativeAdReceived;
}

- (instancetype)initWithConfigLoader:(UIViewController *)viewController configLoader:(AdSwitcherConfigLoader *)configLoader
                            category:(NSString *)category {
    return [self initWithConfigLoader:viewController configLoader:configLoader category:category testMode:NO];
}

- (instancetype)initWithConfigLoader:(UIViewController *)viewController configLoader:(AdSwitcherConfigLoader *)configLoader
                            category:(NSString *)category testMode:(BOOL)testMode {
    
    if (self = [super init]) {
        self.viewController = viewController;
        self.testMode = testMode;

        __block AdSwitcherConfigLoader *configLoaderInBlock = configLoader;
        [configLoader addConfigLoadedHandler:^{
            AdSwitcherConfig *adSwitcherConfig = [configLoaderInBlock adSwitchConfig:category];
            [self initialize:adSwitcherConfig];
        }];
    }
    return self;
}

- (instancetype)initWithConfig:(UIViewController *)viewController config:(AdSwitcherConfig *)adSwitcherConfig {
    return [self initWithConfig:viewController config:adSwitcherConfig testMode:NO];
}

- (instancetype)initWithConfig:(UIViewController *)viewController config:(AdSwitcherConfig *)adSwitcherConfig
                      testMode:(BOOL)testMode {
    
    if (self = [super init]) {
        self.viewController = viewController;
        self.testMode = testMode;

        [self initialize:adSwitcherConfig];
    }
    return self;

}

- (void)setAdReceivedHandler:(void (^)(AdConfig *config, BOOL result))handler {
    nativeAdReceived handlerWrapper = ^(AdConfig *config, BOOL result) {
        dispatch_async(dispatch_get_main_queue(), ^{
            handler(config, result);
        });
    };
    _nativeAdReceived = handlerWrapper;
}

- (AdSwitcherNativeAdData *)getAdData {
    if (!_selectedAdapter) {
        return nil;
    }
    return [_selectedAdapter getAdData];
}

- (void)loadImage:(void (^)(UIImage *uiImage))completion {
    AdSwitcherNativeAdData *adData = [self getAdData];
    if (!adData) {
        completion(nil);
        return;
    }
    [self loadToUIImage:adData.imageUrl completion:completion];
}

- (void)loadIconImage:(void (^)(UIImage *uiImage))completion {
    AdSwitcherNativeAdData *adData = [self getAdData];
    if (!adData) {
        completion(nil);
        return;
    }
    [self loadToUIImage:adData.iconImageUrl completion:completion];
}

- (void)loadToUIImage:(NSString *)imageUrl completion:(void (^)(UIImage *))completion {
    dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
        UIImage *uiImage = [UIImage imageWithData:[NSData dataWithContentsOfURL: [NSURL URLWithString: imageUrl]]];
        dispatch_async(dispatch_get_main_queue(), ^{
            completion(uiImage);
        });
    });
}

- (BOOL)isLoaded {
    return (_selectedAdapter);
}

- (void)openUrl {
    if (!_selectedAdapter) {
        return;
    }
    [_selectedAdapter openUrl];
}

- (void)sendImpression {
    if (!_selectedAdapter) {
        return;
    }
    [_selectedAdapter sendImpression];
}



#pragma - NativeAdDelegate

- (void)nativeAdReceived:(NSObject<AdAdapter> *)adAdapter result:(BOOL)result {
    NSString *className = NSStringFromClass(adAdapter.class);
    _DLOG(@"%@, result=%d", className, result);
    
    if (!_selectedAdConfig || ![_selectedAdConfig.className isEqualToString:className]) {
        _DLOG(@"class name is invalid. selectedClass=%@", _selectedAdConfig.className);
        return;
    }
    
    _loading = false;
    
    if (_nativeAdReceived) {
        _nativeAdReceived([_adConfigDict objectForKey:className], result);
    }
    
    if (!_selectedAdapter) {
        if (result) {
            _selectedAdapter = (NSObject<NativeAdAdapter> *)adAdapter;
        } else {
            [self selectLoad];
        }
    }
}



#pragma - private methods

- (void)selectLoad {
    _loading = YES;
    
    NSObject<NativeAdAdapter> *nativeAdAdapter = nil;
    while (!nativeAdAdapter) {
        _selectedAdConfig = [_adSelector selectAd];
        if (!_selectedAdConfig) {
            break;
        }
        
        [self initAdapter:_selectedAdConfig];
        nativeAdAdapter = [_adapterCacheDict objectForKey:_selectedAdConfig.className];
    }
    
    if (nativeAdAdapter) {
        _DLOG("%@", NSStringFromClass(nativeAdAdapter.class));
        [nativeAdAdapter nativeAdLoad];
        
    } else {
        _loading = NO;
        _DLOG("It will not be able to display all.");
        dispatch_time_t time = dispatch_time(DISPATCH_TIME_NOW, 10 * NSEC_PER_SEC);
        dispatch_after(time, dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
            dispatch_async(dispatch_get_main_queue(), ^{
                [self load];
            });
        });
    }
}

- (void)initAdapter:(AdConfig *)adConfig {
    
    if ([_adapterCacheDict objectForKey:adConfig.className]) {
        return;
    }
    
    Class adAdapterClass = NSClassFromString(adConfig.className);
    if (!adAdapterClass) {
        _DLOG("%@ class is not found.", adConfig.className);
        return;
    }
    
    NSObject<AdAdapter> *adAdapter = [adAdapterClass new];
    
    if (![adAdapter conformsToProtocol:@protocol(NativeAdAdapter)]) {
        _DLOG(@"%@ class is not conform NativeAdAdapter.", adConfig.className);
        return;
    }
    
    NSObject<NativeAdAdapter> *nativeAdAdapter = (NSObject<NativeAdAdapter> *)adAdapter;
    
    @try {
        [nativeAdAdapter setValue:self forKey:@"nativeAdDelegate"];
        
    } @catch (NSException *exception) {
        _DLOG(@"interstitialAdDelegate property is not found. : %@", adConfig.className);
        return;
    }
    
    [nativeAdAdapter nativeAdInitialize:self.viewController parameters:adConfig.parameters testMode:self.testMode];
    
    [_adapterCacheDict setObject:nativeAdAdapter forKey:adConfig.className];
}

- (void)load {
    _DLOG();
    if (_loading) {
        _DLOG("Already started to load");
        return;
    }
    _adSelector = [[AdSelector alloc] initWithConfig:self.adSwitcherConfig];
    [self selectLoad];
}

- (void)initialize:(AdSwitcherConfig *)adSwitcherConfig {
    
    _DLOG("switchType=%d", (int)adSwitcherConfig.switchType);
    
    self.adSwitcherConfig = adSwitcherConfig;
    
    _adapterCacheDict = [NSMutableDictionary<NSString *, NSObject<NativeAdAdapter> *> new];
    
    _adConfigDict = [NSMutableDictionary new];
    for (AdConfig *adConfig in adSwitcherConfig.adConfigArr) {
        [_adConfigDict setObject:adConfig forKey:adConfig.className];
    }
}

@end

File: AdSwitcher-Android/adswitcher/src/main/java/net/adswitcher/AdSwitcherNativeAd.java
package net.adswitcher;

import android.app.Activity;
import android.graphics.drawable.Drawable;
import android.os.Handler;
import android.os.Looper;
import android.util.Log;

import net.adswitcher.adapter.AdAdapter;
import net.adswitcher.adapter.NativeAdAdapter;
import net.adswitcher.adapter.NativeAdListener;
import net.adswitcher.config.AdConfig;
import net.adswitcher.config.AdSwitcherConfig;
import net.adswitcher.config.AdSwitcherConfigLoader;

import java.io.InputStream;
import java.net.HttpURLConnection;
import java.net.URL;
import java.util.HashMap;
import java.util.Map;

/**
 * Created by tkyaji on 2016/10/20.
 */

public class AdSwitcherNativeAd implements NativeAdListener {

    private static final String TAG = "AdSwitcherNativeAd";

    private AdSwitcherNativeAd.AdReceivedListener adReceivedListener;

    private Activity activity;
    private AdSwitcherConfig adSwitcherConfig;
    private boolean testMode;

    private NativeAdAdapter selectedAdapter;
    private AdConfig selectedAdConfig;
    private AdSelector adSelector;
    private Map<String, NativeAdAdapter> adapterCacheMap;
    private Map<String, AdConfig> adConfigMap;
    private boolean loading;

    public AdSwitcherNativeAd(final Activity activity, final AdSwitcherConfigLoader configLoader,
                                final String category) {
        this(activity, configLoader, category, false);
    }

    public AdSwitcherNativeAd(final Activity activity, final AdSwitcherConfigLoader configLoader,
                                final String category, final boolean testMode) {
        this.activity = activity;
        configLoader.addConfigLoadedHandler(new AdSwitcherConfigLoader.ConfigLoadHandler() {
            @Override
            public void onLoaded() {
                final AdSwitcherConfig adSwitcherConfig = configLoader.getAdSwitcherConfig(category);
                activity.runOnUiThread(new Runnable() {
                    @Override
                    public void run() {
                        AdSwitcherNativeAd.this.initialize(adSwitcherConfig, testMode);
                    }
                });
            }
        });
    }

    public AdSwitcherNativeAd(final Activity activity, final AdSwitcherConfig adSwitcherConfig) {
        this(activity, adSwitcherConfig, false);
    }

    public AdSwitcherNativeAd(final Activity activity, final AdSwitcherConfig adSwitcherConfig, final boolean testMode) {
        this.activity = activity;
        this.initialize(adSwitcherConfig, testMode);
    }


    public void load() {
        Log.d(TAG, "load");
        if (this.loading) {
            Log.d(TAG, "Already started to load");
            return;
        }
        this.adSelector = new AdSelector(this.adSwitcherConfig);
        this.selectLoad();
    }

    public AdSwitcherNativeAdData getAdData() {
        if (this.selectedAdapter == null) {
            return null;
        }
        return this.selectedAdapter.getAdData();
    }

    public boolean isLoaded() {
        return this.selectedAdapter != null;
    }

    public void sendImpression() {
        if (this.selectedAdapter == null) {
            return;
        }
        this.selectedAdapter.sendImpression();
    }

    public void openUrl() {
        if (this.selectedAdapter == null) {
            return;
        }
        this.selectedAdapter.openUrl();
    }

    public void loadImage(ImageLoadedListener listener) {
        AdSwitcherNativeAdData adData = this.getAdData();
        if (adData == null) {
            listener.onImageLoaded(null);
        }
        this.loadImageAsync(adData.imageUrl, listener);
    }

    public void loadIconImage(ImageLoadedListener listener) {
        AdSwitcherNativeAdData adData = this.getAdData();
        if (adData == null) {
            listener.onImageLoaded(null);
        }
        this.loadImageAsync(adData.iconImageUrl, listener);
    }

    private void loadImageAsync(final String url, final ImageLoadedListener listener) {
        new Thread(new Runnable() {
            @Override
            public void run() {
                final Drawable drawable = AdSwitcherNativeAd.this.loadImageToDrawable(url);
                AdSwitcherNativeAd.this.activity.runOnUiThread(new Runnable() {
                    @Override
                    public void run() {
                        listener.onImageLoaded(drawable);
                    }
                });
            }
        }).start();
    }

    private Drawable loadImageToDrawable(String url)  {
        try {
            InputStream is = (InputStream)new URL(url).getContent();
            return Drawable.createFromStream(is, "src name");

        } catch (Exception e) {
            Log.d(TAG, e.getMessage(), e);
            return null;
        }
    }


    public void setAdReceivedListener(final AdSwitcherNativeAd.AdReceivedListener listener) {
        this.adReceivedListener = new AdSwitcherNativeAd.AdReceivedListener() {
            @Override
            public void onAdReceived(final AdConfig config, final boolean result) {
                AdSwitcherNativeAd.this.activity.runOnUiThread(new Runnable() {
                    @Override
                    public void run() {
                        listener.onAdReceived(config, result);
                    }
                });
            }
        };
    }


    private void initialize(AdSwitcherConfig adSwitcherConfig, boolean testMode) {
        Log.d(TAG, "testMode=" + testMode);

        this.adSwitcherConfig = adSwitcherConfig;
        this.testMode = testMode;

        this.adapterCacheMap = new HashMap<>();
        this.adConfigMap = new HashMap<>();

        if (this.adSwitcherConfig != null) {
            for (AdConfig config : this.adSwitcherConfig.adConfigList) {
                this.adConfigMap.put(config.className, config);
            }
        }
    }

    private void selectLoad() {
        this.loading = true;

        NativeAdAdapter nativeAdAdapter = null;
        while (nativeAdAdapter == null) {
            this.selectedAdConfig = this.adSelector.selectAd();
            if (this.selectedAdConfig == null) {
                break;
            }

            this.initAdapter(this.selectedAdConfig);
            nativeAdAdapter = this.adapterCacheMap.get(this.selectedAdConfig.className);
        }

        if (nativeAdAdapter != null) {
            Log.d(TAG, nativeAdAdapter.getClass().getName());
            final NativeAdAdapter _nativeAdAdapter = nativeAdAdapter;
            AdSwitcherNativeAd.this.activity.runOnUiThread(new Runnable() {
                @Override
                public void run() {
                    _nativeAdAdapter.nativeAdLoad();
                }
            });

        } else {
            this.loading = false;
            Log.w(TAG, "It will not be able to display all.");

            (new Handler()).postDelayed(new Runnable() {
                @Override
                public void run() {
                    AdSwitcherNativeAd.this.activity.runOnUiThread(new Runnable() {
                        @Override
                        public void run() {
                            AdSwitcherNativeAd.this.load();
                        }
                    });
                }
            }, 10000);
        }
    }

    private void initAdapter(final AdConfig adConfig) {
        if (this.adapterCacheMap.containsKey(adConfig.className)) {
            return;
        }

        try {
            Class<?> adAdapterClass = Class.forName(adConfig.className);
            final NativeAdAdapter nativeAdAdapter = (NativeAdAdapter) adAdapterClass.newInstance();

            this.activity.runOnUiThread(new Runnable() {
                @Override
                public void run() {
                    nativeAdAdapter.nativeAdInitialize(AdSwitcherNativeAd.this.activity,
                            AdSwitcherNativeAd.this, adConfig.parameters,
                            AdSwitcherNativeAd.this.testMode);
                }
            });
            this.adapterCacheMap.put(adConfig.className, nativeAdAdapter);

        } catch (ClassNotFoundException ex) {
            Log.w(TAG, adConfig.className + " class is not found.");
            return;

        } catch (Throwable e) {
            Log.w(TAG, adConfig.className + " class is not conform NativeAdAdapter.");
            return;
        }
    }


    @Override
    public void nativeAdReceived(AdAdapter adAdapter, boolean result) {
        String className = adAdapter.getClass().getName();
        Log.d(TAG, "nativeAdReceived : " + className + ", result=" + result);

        if (this.selectedAdConfig == null) {
            Log.d(TAG, "class name is invalid.(null)");
            return;
        } else if (!this.selectedAdConfig.className.equals(className)) {
            Log.d(TAG, "class name is invalid. selectedClass=" + this.selectedAdConfig.className);
            return;
        }

        this.loading = false;

        if (this.selectedAdapter == null && result) {
            this.selectedAdapter = (NativeAdAdapter) adAdapter;
        }

        if (this.adReceivedListener != null) {
            this.adReceivedListener.onAdReceived(this.adConfigMap.get(className), result);
        }

        if (this.selectedAdapter == null) {
            if (!result) {
                this.selectLoad();
            }
        }
    }



    public interface AdReceivedListener {
        void onAdReceived(AdConfig config, boolean result);
    }

    public interface ImageLoadedListener {
        void onImageLoaded(Drawable drawable);
    }
}
